File: src/game/object/Coin.java
package game.object;

import engine.Camera;
import game.Assets;

public class Coin {
	private static final float SIZE = 32.0f;

	float x;
	float y;

	public CoinLevel tier;
	public boolean collected = false;

	public Coin(float gx, float gy, CoinLevel type) {
		this.x = gx;
		this.y = gy;
		this.tier = type;
	}

	public double distTo(double px, double py) {
		return Math.sqrt(Math.pow(px - x, 2.0) + Math.pow(py - y, 2.0));
	}

	public void render(Camera camera) {
		if (collected) return;

		var texture = tier == CoinLevel.gold ? Assets.coinTexture
			: tier == CoinLevel.heavy ? Assets.blueCoinTexture
			: tier == CoinLevel.delta ? Assets.superCoinTexture
			: Assets.keyCoinTexture;

		texture.bind();
		Assets.textureShader.enable().setMVP(camera.getMVPCentered(
			x,
			y,
			SIZE,
			SIZE
		));
		Assets.rect.render();
	}
}

File: src/game/object/Coin.kt
package game.`object`

import engine.Camera
import engine.Texture
import game.Assets
import kotlin.math.sqrt

class Coin(var x: Float, var y: Float, var tier: CoinLevel) {
	var collected = false

	fun distTo(px: Float, py: Float) = sqrt((px - x) * (px - x) + (py - y) * (py - y))

	fun render(camera: Camera) {
		if (collected) return
		val texture: Texture = when (tier) {
			CoinLevel.GOLD -> Assets.coinTexture
			CoinLevel.HEAVY -> Assets.blueCoinTexture
			CoinLevel.DELTA -> Assets.superCoinTexture
			else -> Assets.keyCoinTexture
		}
		texture.bind()
		Assets.textureShader.enable().setMVP(
			camera.getMVPCentered(
				x,
				y,
				SIZE,
				SIZE
			)
		)
		Assets.rect.render()
	}

	companion object {
		private const val SIZE = 32.0f
	}
}
